src/lib.rs

```diff
@@ -355,8 +378,10 @@ macro_rules! py_run_impl {
         use $crate::types::IntoPyDict;
         use $crate::ToPyObject;
         let d = [$((stringify!($val), $val.to_object($py)),)+].into_py_dict($py);
-
-        if let Err(e) = $py.run($code, None, Some(d)) {
+        $crate::py_run_impl!($py, *d, $code)
```

Comment:
What do you think of making a `#[doc(hidden)]` module `pyo3::impl_` and moving `py_run_impl` into there?

I think we could also move `pyo3::impl_::derive_utils` and `pyo3::impl_::internal_tricks` into such a module.

This way it'd be easier for users to see what bits of pyo3 are considered internals that they should not depend upon!

Reply:
Hmm 🤔, I think there are advantages and disadvantages to doing so.
Placing here would be clearer for someone who wants to read the implementation.